esp-hal/src/spi/master.rs

```diff
@@ -1535,38 +1551,107 @@ pub mod dma {
         }
 
         pub fn transfer_in_place(&mut self, words: &mut [u8]) -> Result<(), Error> {
-            let mut spi_dma = self.spi_dma.take().unwrap();
-            let (mut tx_buf, mut rx_buf) = self.buffers.take().unwrap();
+            let (mut spi_dma, mut tx_buf, mut rx_buf) = self.wait_for_idle();
 
             let chunk_size = min(tx_buf.capacity(), rx_buf.capacity());
 
             for chunk in words.chunks_mut(chunk_size) {
                 tx_buf.fill(chunk);
                 rx_buf.set_length(chunk.len());
 
-                let transfer = match spi_dma.dma_transfer(tx_buf, rx_buf) {
-                    Ok(transfer) => transfer,
+                match spi_dma.dma_transfer(tx_buf, rx_buf) {
+                    Ok(transfer) => self.state = State::Transferring(transfer),
                     Err((e, spi, tx, rx)) => {
-                        self.spi_dma = Some(spi);
-                        self.buffers = Some((tx, rx));
+                        self.state = State::Idle(spi, tx, rx);
+                        return Err(e);
+                    }
+                }
+                (spi_dma, tx_buf, rx_buf) = self.wait_for_idle();
+
+                let bytes_read = rx_buf.read_received_data(chunk);
+                debug_assert_eq!(bytes_read, chunk.len());
+            }
+
+            self.state = State::Idle(spi_dma, tx_buf, rx_buf);
+            Ok(())
+        }
+    }
+
+    impl<'d, T, C, D, M> HalfDuplexReadWrite for SpiDmaBus<'d, T, C, D, M>
+    where
+        T: InstanceDma,
+        C: DmaChannel,
+        C::P: SpiPeripheral,
+        D: IsHalfDuplex,
+        M: Mode,
+    {
+        type Error = super::Error;
+
+        /// Half-duplex read.
+        fn read(
+            &mut self,
+            data_mode: SpiDataMode,
+            cmd: Command,
+            address: Address,
+            dummy: u8,
+            buffer: &mut [u8],
+        ) -> Result<(), Self::Error> {
+            let (mut spi_dma, mut tx_buf, mut rx_buf) = self.wait_for_idle();
+
+            for chunk in buffer.chunks_mut(rx_buf.capacity()) {
+                rx_buf.set_length(chunk.len());
+
+                match spi_dma.read(data_mode, cmd, address, dummy, rx_buf) {
```

Comment:
This isn't quite right I think, you only want to do the `cmd`, `address` and `dummy` phase in the first iteration of the loop.

My simple plan for this was to limit `buffer.len()` to be at most `rx_buf.capacity()` but I'm not sure if that's too limiting.

Reply:
Good point! I believe if we set `Command`, `Address` etc to `None` the API after the first iteration it should skip sending those again and just pump out the data. But for now I think your option is a nice simple one, and it's how it worked before.